src/mugs.js

```diff
@@ -243,36 +340,120 @@ define([
+                var obj = messages[i];
+                if (obj.key === msg.key) {
+                    if (obj.level === msg.level && obj.message === msg.message) {
+                        // added messages already exist (no change)
+                        return false;
+                    }
+                    messages.splice(i, 1);
+                    removed = true;
+                    break;
+                }
+            }
+            if (msg.message) {
+                messages.push(msg);
+            } else if (!removed) {
+                return false;
             }
+            if (messages.length) {
+                this.messages[attr] = messages;
+            } else {
+                delete this.messages[attr];
+            }
+            return true;
+        },
+        get: function (attr) {
+            if (arguments.length) {
+                return _.flatten(_.pluck(this.messages[attr || ""], "message"));
+            }
+            return _.flatten(_.map(this.messages, function (messages) {
+                return _.pluck(messages, "message");
+            }), _.identity);
+        },
+        /**
+         * Execute a function for each message
+         *
+         * @param attr - Optional property limiting the messages visited.
+         *          The callback signature is `callback(msg)` if
+         *          this argument is provided, and otherwise
+         *          `callback(msg, property)`. In all cases the first
+         *          argument `msg` is a message object.
+         * @param callback - A function to be called for each message object.
+         */
+        each: function () {
+            var attr, callback;
+            if (arguments.length > 1) {
+                attr = arguments[0] || "";
+                callback = arguments[1];
+                _.each(this.messages[attr], callback);
+            } else {
+                callback = arguments[0];
+                _.each(this.messages, function (messages, attr) {
+                    _.each(messages, function (msg) {
+                        callback(msg, attr);
+                    });
+                });
+            }
+        },
+        toString: function () {
+            var messages = [],
+                last = null;
+            this.each(function (msg, attr) {
+                if (attr !== last) {
+                    messages.push(attr + ":");
+                    last = attr;
+                }
+                messages.push("  - " + msg.message); // + " [" + msg.key + "]");
+            });
+            return messages.join("\n");
         }
-
-        return retBlock;
-    }
+    };
 
     function MugProperties (options) {
         this.__data = {};
         this.__spec = options.spec;
         this.__mug = options.mug;
-        this.shouldChange = options.shouldChange || function () { return function () {}; };
+        this.shouldChange = function () { return function () {}; };
```

Comment:
The interface for `shouldChange` is documented on the form object: https://github.com/dimagi/Vellum/blob/mug-messages/src/form.js#L696-L697